std/json.c3

```diff
@@ -0,0 +1,428 @@
+    // TODO this should be const (can't use with has_value() right now).
+    let next_char : Optional<u8> = cursor:peak();
+
+    let next_char_data : u8 = next_char:data();
+    if next_char:has_value() and next_char_data == token {
+        cursor:next();
+        return true;
+    }
+
+    return false;
+}
+
+function eat : (cursor : Cursor&, string : StringView) -> bool = {
+    for token in string:get_iter() {
+        if !cursor:eat(token) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+function eat_whitespace : (cursor: Cursor&) -> void = {
+    while !cursor:is_eof() {
+        let next_char : Optional<u8> = cursor:peak();
+        runtime_assert(next_char:has_value());
+
+        if !is_json_whitespace(next_char:data()) {
+            return;
+        }
+
+        // Eat the current character.
+        cursor:next();
+    }
+}
+
+// TODO return ErrorOr<JSON_Node>.
+function parse_object : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    if !cursor:eat(char("{")) {
+        return make_optional<JSON_Node>();
+    }
+
+    let node : JSON_Node = new_json_node(/* object */ 6);
+
+    cursor:eat_whitespace();
+
+    if cursor:eat(char("}")) {
+        // Empty object.
+        return make_optional(node);
+    }
+
+    // Initialize children vector.
+    node.children:store(cursor.allocator:make_vector<JSON_Node>());
+
+    while true {
+        cursor:eat_whitespace();
+
+        let string_node : Optional<JSON_Node> = cursor:parse_string();
+
+        if !string_node:has_value() {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+
+        node.children:data():push_back(string_node:data());
+
+        cursor:eat_whitespace();
+
+        if !cursor:eat(char(":")) {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+
+        let value_node : Optional<JSON_Node> = cursor:parse_element();
+
+        if !value_node:has_value() {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+
+        node.children:data():push_back(value_node:data());
+
+        if cursor:eat(char("}")) {
+            // We're done.
+            return make_optional(node);
+        }
+
+        if !cursor:eat(char(",")) {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+    }
+
+    // Unreachable.
+    return make_optional(node);
+}
+
+function parse_array : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    if !cursor:eat(char("[")) {
+        return make_optional<JSON_Node>();
+    }
+
+    let node : JSON_Node = new_json_node(/* array */ 5);
+
+    cursor:eat_whitespace();
+
+    if cursor:eat(char("]")) {
+        // Empty array.
+        return make_optional(node);
+    }
+
+    // Initialize children vector.
+    node.children:store(cursor.allocator:make_vector<JSON_Node>());
```

Comment:
Same comment here about the using a length 0 children vector instead of an optional